Move CapRegState into a new cap_register_t field

arichardson opened the pull request:
This allows us to simply copy cap_register_t values and not have to worry
about updating the external capreg_state array.
Longer-term it would be good to have cheri-compressed-cap transparently
handle the bounds decoded/raw bits states, but for now it is a lot easier
to add a cr_extra field to retain the existing meaning of the CREG_*
constants. This also means we don't have to adjust the Morello TCG code
that uses these constants.

arichardson committed:
[Morello] Use the helper functions to create NULL/max perm capabilities

While the existing code currently works, I have some refactoring changes
to move the capreg state into the cap_register_t. With that follow-up
change the capreg state will not be initialized by the current code so
all special Morello capability registers are treated as integers which
causes assertions later on.

arichardson committed:
[Morello][NFC] Avoid direct accesses to env->pc.cap

Use the helper functions instead in preparation for added debug assertions.
While touching those lines also make checkpatch.pl happy by adding braces.

arichardson committed:
Move CapRegState into a new cap_register_t field

This allows us to simply copy cap_register_t values and not have to worry
about updating the external capreg_state array.
Longer-term it would be good to have cheri-compressed-cap transparently
handle the bounds decoded/raw bits states, but for now it is a lot easier
to add a cr_extra field to retain the existing meaning of the CREG_*
constants. This also means we don't have to adjust the Morello TCG code
that uses these constants.

arichardson committed:
[Morello] Add a load_and_decompress_cap_from_memory_raw() helper

This avoids some duplicated code in the load+branch helpers. It is also
required for upcoming refactoring that asserts valid CapRegState for
jump targets.

arichardson commented:
No significant performance difference according to perf stat.

After:
```

 Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh -device virtio-rng-pci' (5 runs):

       7869.069734      task-clock (msec)         #    0.991 CPUs utilized            ( +-  0.76% )
             2,267      context-switches          #    0.288 K/sec                    ( +-  0.54% )
                 1      cpu-migrations            #    0.000 K/sec                    ( +- 66.67% )
            12,945      page-faults               #    0.002 M/sec                    ( +-  0.28% )
    25,997,037,004      cycles                    #    3.304 GHz                      ( +-  0.33% )
    70,627,655,187      instructions              #    2.72  insn per cycle           ( +-  0.06% )
    10,672,456,927      branches                  # 1356.254 M/sec                    ( +-  0.07% )
       163,178,624      branch-misses             #    1.53% of all branches          ( +-  0.20% )

       7.936581201 seconds time elapsed                                          ( +-  0.72% )
```
Before:
```
 Performance counter stats for '/local/scratch/alr48/cheri/output/sdk/bin/qemu-system-riscv64cheri.v5.2.0-1215-g6f94b82947 -M virt -m 2048 -nographic -bios bbl-riscv64cheri-virt-fw_jump.bin -kernel /local/scratch/alr48/cheri/output/kernel-riscv64-purecap.CHERI-PURECAP-QEMU-MFS-ROOT -append init_path=/sbin/startup-benchmark.sh -device virtio-rng-pci' (5 runs):

       7864.894783      task-clock (msec)         #    0.965 CPUs utilized            ( +-  1.10% )
             2,271      context-switches          #    0.289 K/sec                    ( +-  1.99% )
                 2      cpu-migrations            #    0.000 K/sec                    ( +- 42.39% )
            12,786      page-faults               #    0.002 M/sec                    ( +-  0.81% )
    25,870,861,867      cycles                    #    3.289 GHz                      ( +-  0.27% )
    70,282,360,955      instructions              #    2.72  insn per cycle           ( +-  0.13% )
    10,670,899,986      branches                  # 1356.776 M/sec                    ( +-  0.12% )
       163,127,585      branch-misses             #    1.53% of all branches          ( +-  0.22% )

       8.149516958 seconds time elapsed                                          ( +-  2.76% )
```

Slightly more cycles, but that is probably just due to different LTO inlining choices or similar.

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -1061,7 +1067,7 @@ static inline void gen_move_cap_gp_select_gp(DisasContext *ctx, int dest_num,
                              tcg_cond_string(cond), true_source_num,
                              false_source_num);
-
+    /* FIXME: Why do we require that the register is decompressed here? */
```
Presumably because not doing so complicates the TCG-level tracking of compression state

arichardson replied:
Well with this change it should no longer be required. @LawrenceEsswood is that correct?

jrtc27 replied:
TCG uses it for optimisation purposes, to avoid redundant decompression, done at translate time

LawrenceEsswood replied:
This, somewhat. The tracking would be perfectly happy to have a union of possible states for a conditional move if it were a better choice. AFAICR, I decided that any move should decompress. This was mostly because move was being used for basically every instruction that had a capability write (which are implemented as a move, and then an in-place operation). I thought that if we were going to move a register, it was a good idea to decompress before rather than after (as it was likely going to be decompressed anyway). That way, the same amount of work was being done but leaving more things decompressed. My thinking was, if a register is being moved around we will probably at some point use its bounds.

LawrenceEsswood replied:
Although, at the time, way more fields required decompression (type, perms etc). Maybe this is no longer as true.

LawrenceEsswood replied:
I guarantee some code will break if we remove it though. Although the get_base / get_top etc family ensure decompression, they do so via a 'gen_ensure_cap_decompressed', which kills temps. If we ever want to get rid of gen_ensure_cap_decompressed in move, lots of callers will likely have to do a gen_ensure_cap_decompressed just before the move anyway, before any useful temps are created.

arichardson replied:
> TCG uses it for optimisation purposes, to avoid redundant decompression, done at translate time

Yes but we could just copy the translation context capreg state from the other register?

arichardson replied:
Although this is the conditional one so maybe a bit more awkward

arichardson replied:
But you are right that this tcg state tracking makes a difference, in the initial version of this patch I removed the `gen_lazy_cap_set_state(ctx, dest_num, CREG_FULLY_DECOMPRESSED);` below instead of updating it and immediately got some tcg_aborts due to dead temps.

jrtc27 commented:
> Longer-term it would be good to have cheri-compressed-cap transparently
> handle the bounds decoded/raw bits states

I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...

jrtc27 commented:
> > Longer-term it would be good to have cheri-compressed-cap transparently
> > handle the bounds decoded/raw bits states
> 
> I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...

(and QEMU needs to track it itself for TCG purposes anyway, so it wouldn't be transparent)

jrtc27 commented on target/cheri-common/op_helper_cheri_common.c:
```diff
@@ -1336,6 +1336,18 @@ bool load_cap_from_memory_raw(CPUArchState *env, target_ulong *pesbt,
                                         retpc, physaddr, NULL);
 }
 
+cap_register_t load_and_decompress_cap_from_memory(
```
With a _raw suffix? The non-raw load_cap_from_memory also updates the compressed state of cd, so the name here would imply this takes cd and does the same, but it doesn't.

arichardson replied:
Yeah I guess I could add a _raw suffix, I thought it isn't necessary since we are returning the cap_register_t and therefore aren't updating any registers.

arichardson commented:
> > Longer-term it would be good to have cheri-compressed-cap transparently
> > handle the bounds decoded/raw bits states
> 
> I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...

I used uint8_t since we already have three other uint8_t fields. Regarding the toher comment, I was thinking of changing cheri-compressed-cap to set a "bounds have been decoded" flag but I guess we can always do that in qemu. The reason I was thinking of moving it to cheri-compressed-cap is that there were various places in QEMU where that wasn't being done and I had to go and find them.

arichardson commented:
> > > Longer-term it would be good to have cheri-compressed-cap transparently
> > > handle the bounds decoded/raw bits states
> > 
> > 
> > I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...
> 
> I used uint8_t since we already have three other uint8_t fields. Regarding the toher comment, I was thinking of changing cheri-compressed-cap to set a "bounds have been decoded" flag but I guess we can always do that in qemu. The reason I was thinking of moving it to cheri-compressed-cap is that there were various places in QEMU where that wasn't being done and I had to go and find them.

And I agree that cheri-compressed-cap should not have something like the untagged_cap/tagged_cap states just a few helpers to lazily handle bounds decompression.

LawrenceEsswood commented:
> > > > Longer-term it would be good to have cheri-compressed-cap transparently
> > > > handle the bounds decoded/raw bits states
> > > 
> > > 
> > > I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...
> > 
> > 
> > I used uint8_t since we already have three other uint8_t fields. Regarding the toher comment, I was thinking of changing cheri-compressed-cap to set a "bounds have been decoded" flag but I guess we can always do that in qemu. The reason I was thinking of moving it to cheri-compressed-cap is that there were various places in QEMU where that wasn't being done and I had to go and find them.
> 
> And I agree that cheri-compressed-cap should not have something like the untagged_cap/tagged_cap states just a few helpers to lazily handle bounds decompression.

TBH, I am not sure that QEMU wants the the current lazy states as is. I think we should just have states be integer, compressed, and decompressed. And then always have cursor/pesbt/tag be correct, and other fields valid only if decompressed. There is currently loads of rubbish in getting the tag where the logic is (state == tagged) || (state == decompressed && tag), which really should just be (tag). But I also agree with Alex. We should hide internal fields, create accessors, and have internal caching of decompression state handled by the library. I think it is, at worst, no different for other consumers of the compression library. And at best, it is more performant.

arichardson commented:
> > > > > Longer-term it would be good to have cheri-compressed-cap transparently
> > > > > handle the bounds decoded/raw bits states
> > > > 
> > > > 
> > > > I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...
> > > 
> > > 
> > > I used uint8_t since we already have three other uint8_t fields. Regarding the toher comment, I was thinking of changing cheri-compressed-cap to set a "bounds have been decoded" flag but I guess we can always do that in qemu. The reason I was thinking of moving it to cheri-compressed-cap is that there were various places in QEMU where that wasn't being done and I had to go and find them.
> > 
> > 
> > And I agree that cheri-compressed-cap should not have something like the untagged_cap/tagged_cap states just a few helpers to lazily handle bounds decompression.
> 
> TBH, I am not sure that QEMU wants the the current lazy states as is. I think we should just have states be integer, compressed, and decompressed. And then always have cursor/pesbt/tag be correct, and other fields valid only if decompressed. There is currently loads of rubbish in getting the tag where the logic is (state == tagged) || (state == decompressed && tag), which really should just be (tag). But I also agree with Alex. We should hide internal fields, create accessors, and have internal caching of decompression state handled by the library. I think it is, at worst, no different for other consumers of the compression library. And at best, it is more performant.

If we could use C++ we could mark the bounds fields as private and require use of the member functions... I guess we can rename them to `_dont_access_this_directly_cr_base` or similar.

jrtc27 commented:
> > > > > > Longer-term it would be good to have cheri-compressed-cap transparently
> > > > > > handle the bounds decoded/raw bits states
> > > > > 
> > > > > 
> > > > > I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...
> > > > 
> > > > 
> > > > I used uint8_t since we already have three other uint8_t fields. Regarding the toher comment, I was thinking of changing cheri-compressed-cap to set a "bounds have been decoded" flag but I guess we can always do that in qemu. The reason I was thinking of moving it to cheri-compressed-cap is that there were various places in QEMU where that wasn't being done and I had to go and find them.
> > > 
> > > 
> > > And I agree that cheri-compressed-cap should not have something like the untagged_cap/tagged_cap states just a few helpers to lazily handle bounds decompression.
> > 
> > 
> > TBH, I am not sure that QEMU wants the the current lazy states as is. I think we should just have states be integer, compressed, and decompressed. And then always have cursor/pesbt/tag be correct, and other fields valid only if decompressed. There is currently loads of rubbish in getting the tag where the logic is (state == tagged) || (state == decompressed && tag), which really should just be (tag). But I also agree with Alex. We should hide internal fields, create accessors, and have internal caching of decompression state handled by the library. I think it is, at worst, no different for other consumers of the compression library. And at best, it is more performant.
> 
> If we could use C++ we could mark the bounds fields as private and require use of the member functions... I guess we can rename them to `_dont_access_this_directly_cr_base` or similar.

That's what the underscore convention is meant to be for...

arichardson commented:
I also think having another flag for tagged+unsealed would be useful, the bitshifts from cap_is_unsealed() do show up in perf traces.

arichardson commented:
> > > > > > > Longer-term it would be good to have cheri-compressed-cap transparently
> > > > > > > handle the bounds decoded/raw bits states
> > > > > > 
> > > > > > 
> > > > > > I disagree; cheri-compressed-cap is for dealing with the compression scheme, not a "do everything exactly how QEMU wants" library. Different consumers will have different requirements. Having cr_extra is probably ok, though uint8_t is a bit arbitrary and limited...
> > > > > 
> > > > > 
> > > > > I used uint8_t since we already have three other uint8_t fields. Regarding the toher comment, I was thinking of changing cheri-compressed-cap to set a "bounds have been decoded" flag but I guess we can always do that in qemu. The reason I was thinking of moving it to cheri-compressed-cap is that there were various places in QEMU where that wasn't being done and I had to go and find them.
> > > > 
> > > > 
> > > > And I agree that cheri-compressed-cap should not have something like the untagged_cap/tagged_cap states just a few helpers to lazily handle bounds decompression.
> > > 
> > > 
> > > TBH, I am not sure that QEMU wants the the current lazy states as is. I think we should just have states be integer, compressed, and decompressed. And then always have cursor/pesbt/tag be correct, and other fields valid only if decompressed. There is currently loads of rubbish in getting the tag where the logic is (state == tagged) || (state == decompressed && tag), which really should just be (tag). But I also agree with Alex. We should hide internal fields, create accessors, and have internal caching of decompression state handled by the library. I think it is, at worst, no different for other consumers of the compression library. And at best, it is more performant.
> > 
> > 
> > If we could use C++ we could mark the bounds fields as private and require use of the member functions... I guess we can rename them to `_dont_access_this_directly_cr_base` or similar.
> 
> That's what the underscore convention is meant to be for...

Yeah except that we (I) started to ignore that when convenient...

LawrenceEsswood commented:
On 2021-10-15 16:20, Alexander Richardson wrote:
>> Longer-term it would be good to have cheri-compressed-cap
>> transparently
>> handle the bounds decoded/raw bits states
>> 
>> I disagree; cheri-compressed-cap is for dealing with the compression
>> scheme, not a "do everything exactly how QEMU wants" library.
>> Different consumers will have different requirements. Having
>> cr_extra is probably ok, though uint8_t is a bit arbitrary and
>> limited...
> 
> I used uint8_t since we already have three other uint8_t fields.
> Regarding the toher comment, I was thinking of changing
> cheri-compressed-cap to set a "bounds have been decoded" flag but I
> guess we can always do that in qemu. The reason I was thinking of
> moving it to cheri-compressed-cap is that there were various places in
> QEMU where that wasn't being done and I had to go and find them.
> 
> And I agree that cheri-compressed-cap should not have something like
> the untagged_cap/tagged_cap states just a few helpers to lazily handle
> bounds decompression.
> 
> TBH, I am not sure that QEMU wants the the current lazy states as is.
> I think we should just have states be integer, compressed, and
> decompressed. And then always have cursor/pesbt/tag be correct, and
> other fields valid only if decompressed. There is currently loads of
> rubbish in getting the tag where the logic is (state == tagged) ||
> (state == decompressed && tag), which really should just be (tag). But
> I also agree with Alex. We should hide internal fields, create
> accessors, and have internal caching of decompression state handled by
> the library. I think it is, at worst, no different for other consumers
> of the compression library. And at best, it is more performant.
> 
> If we could use C++ we could mark the bounds fields as private and
> require use of the member functions... I guess we can rename them to
> _dont_access_this_directly_cr_base or similar.
> 

The thing is, modifying pesbt/cursor also wants protecting, as we need 
to mark internal state to being no longer decompressed. I don't think 
its important, but I think the compression library should be hiding 
decompression caching.

> --
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub [1], or unsubscribe
> [2].
> Triage notifications on the go with GitHub Mobile for iOS [3] or
> Android [4].
> 
> Links:
> ------
> [1] https://github.com/CTSRD-CHERI/qemu/pull/181#issuecomment-944387125
> [2] 
> https://github.com/notifications/unsubscribe-auth/ABSDMFU6ZOOJREQ55NDIXFDUHBBCZANCNFSM5GBVAOBQ
> [3] 
> https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&amp;mt=8&amp;pt=524675
> [4] 
> https://play.google.com/store/apps/details?id=com.github.android&amp;referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub

LawrenceEsswood commented:
Anyway, this patch looks fine to me, but Jenkins is getting some asserts?

arichardson commented:
> Anyway, this patch looks fine to me, but Jenkins is getting some asserts?

Looks like it's related to helper_exception_return. Will fix.

arichardson committed:
[Morello] Ensure that all special capability registers are canonical

An all-zeroes cap_register_t is not a canonical representation of a null
capability and can result in assertions when copied since capreg state
is assumed to be that of an integer but pesbt is not set to NULL pesbt.

This change is probably not quite correct since there are some fields that
are capabilities but marked as ARM_CP_ALIAS (although I can can't see
what they are aliased to).

arichardson commented:
Should be fixed now, but I'm not convinced the new capability init logic is quite correct. @LawrenceEsswood could you take a look.

jrtc27 commented on target/arm/cpu.c:
```diff
@@ -153,13 +175,7 @@ static void cp_reg_reset(gpointer key, gpointer value, gpointer opaque)
-    } else
-#endif
-        if (cpreg_field_is_64bit(ri)) {
+    if (cpreg_field_is_64bit(ri)) {
```
You might want to have an empty if before this for if has_special_capresetvalue is true? Otherwise if resetvalue doesn't match capresetvalue's cursor things will get weird.

arichardson replied:
Good catch!

jrtc27 commented on target/arm/cpu.c:
```diff
@@ -153,13 +174,9 @@ static void cp_reg_reset(gpointer key, gpointer value, gpointer opaque)
-    } else
-#endif
-        if (cpreg_field_is_64bit(ri)) {
+        /* Initialized above. */
```
Worth an assertion this is never hit ifndef TARGET_CHERI? Maybe it's ok given the function is defined to return the constant false in that case? That or could (slightly unnecessarily) keep this ifdef'ed.

arichardson replied:
Could add the assert if you think it's useful, but I feel it should be obvious that capregs don't exist for non-CHERI. Removing the ifdef has the added benefit of working around the clang-format weirdness.

jrtc27 replied:
Yeah, probably not worth it

LawrenceEsswood commented on target/arm/cpu.c:
```diff
@@ -136,6 +136,27 @@ static void cp_reg_reset(gpointer key, gpointer value, gpointer opaque)
+         * as pesbt and capreg_state are initialized correctly.
+         * The default reset value is null unless has_special_capresetvalue is
+         * set.
+         * TODO: We do this even for (ARM_CP_SPECIAL | ARM_CP_ALIAS) since
```
Comment says TODO, but it is done. I wonder if this causes any problems? I don't think it does.

LawrenceEsswood replied:
I am a little concerned about ARM_CP_ALIAS not being ignored here. My worry is that an aliasing register will use an offset into another register, and resetting both in the wrong order could do something funny. I had a scan through and it doesn't look like there are any cap aliases that could cause this problem.

arichardson replied:
For some reason it appears that some registers are tagged as aliases even though I can't see an aliasee.

arichardson replied:
```
CPREG_CAP_INIT: VBAR_EL2_S  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL1  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDR_EL2_S  ARM_CP_ALIAS
CPREG_CAP_INIT: SP_EL0  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL2  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL12  ARM_CP_ALIAS
CPREG_CAP_INIT: SP_EL1  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL3  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR_EL12  ARM_CP_ALIAS
CPREG_CAP_INIT: SP_EL2  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR_S  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDRURW_S  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDRURO_S  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDRPRW_S  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR_EL2_S  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL1  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDR_EL2_S  ARM_CP_ALIAS
CPREG_CAP_INIT: SP_EL0  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL2  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL12  ARM_CP_ALIAS
CPREG_CAP_INIT: SP_EL1  ARM_CP_ALIAS
CPREG_CAP_INIT: ELR_EL3  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR_EL12  ARM_CP_ALIAS
CPREG_CAP_INIT: SP_EL2  ARM_CP_ALIAS
CPREG_CAP_INIT: VBAR_S  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDRURW_S  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDRURO_S  ARM_CP_ALIAS
CPREG_CAP_INIT: TPIDRPRW_S  ARM_CP_ALIAS
```

arichardson replied:
Fixed in the latest version.

LawrenceEsswood commented on target/arm/cpu.c:
```diff
@@ -136,6 +136,27 @@ static void cp_reg_reset(gpointer key, gpointer value, gpointer opaque)
 
+#ifdef TARGET_CHERI
+    if (cpreg_field_is_cap(ri)) {
+        assert(ri->fieldoffset && "cpreg cap without offset?");
```
Won't be hit yet, but an offset of zero is expected for any ARM_CP_SPECIAL that are also caps. Currently none are handled that way, but its allowed. Possibly the if (ri->type & ARM_CP_SPECIAL) should really be pulled before this if block to allow that to work? Or at least only throw this assert if not ARM_CP_SPECIAL?

arichardson replied:
I added the assertion to catch future changes and assert on startup.

LawrenceEsswood reviewed (approved):
The change regarding ARM_CP_SPECIAL and ARM_CP_ALIAS might cause a problem in the future if somebody creates a specially handled cap register, or creates a integer register that aliases with a part of a cap register. But it doesn't look wrong as is.

arichardson committed:
[Morello] Initialize ARM_CP_ALIAS capregs explicitly in arm_cpu_reset

This allows us to no longer reset (ARM_CP_SPECIAL | ARM_CP_ALIAS) registers
in cp_reg_reset(). This should avoid some additional work on reset and
missing initializations will be caught by the assertions in
cp_reg_check_reset().

arichardson committed:
[Morello] Use an early return to reduce diff to upstream.
